[TCFMM-12] Initialize ticks with "infinite" index at origination

rinn7e opened the pull request:
## Description

Problem: Following the discussion in which the tick initialization is dependent
on existing "witnesses" ticks, there needs to be the two "extreme" ticks
(highest and lowest allowed indexes) at origination.

We need to pick a large number to represent this "infinite" index.

Solution: Add `default_storage` which is mean to be used at the
origination, add the initialized `ticks` with infinite index.

<!--
Describes the nature of your changes. If they are substantial, you should
further subdivide this into a section describing the problem you are solving and
another describing your solution.
-->

[//]: # (Here you can add a link to the corresponding issue tracker, e. g. https://issues.serokell.io/issue/AD-)
[//]: # (For GitHub/GitLab issues it is better to use just hash symbol or exclamation mark as it is more resistant to repo movements)
[//]: # (In this case please also prefix the link with "Resolves" keyword)
[//]: # (See https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
[//]: # (See https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords)
## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

Resolves TCFMM-12

Depends on #6 
## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

[//]: # (Update link to style guide if necesary or remove if it's not present)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).
Problem: `liquidity_net` (old `liquidity_delta`) is currently always
subtracted, that does not seem right.

Solution: when we exit a tick while going up, add `liquidity_net`
instead of subtracting it.

rinn7e commented:
@dcastro @Martoon-00 Took me a while to figure out what need to be done.

Basically, I add a make command to create a `default_storage.tz` with infinity index, currently it looks like this:

```
(Pair (Pair (Pair (Pair (Pair 0 0) 0 0) 0 "1970-01-01T00:00:00Z")
            (Pair 0
                  -10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)
            {}
            {})
      (Pair 0 0)
      { Elt -10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
            (Pair (Pair (Pair (Pair 0 0) 0)
                        0
                        10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)
                  (Pair -10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
                        0)
                  0
                  0) ;
        Elt 10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
            (Pair (Pair (Pair (Pair 0 0) 0)
                        0
                        10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)
                  (Pair -10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
                        0)
                  0
                  0) }
      0)

```

I only focus on the `ticks` field for now, the other generated field value may be incorrect/ not desired at the moments. Hope to get your feedback on this.

rinn7e commented:
> One more thing is that IMO we should add the default storage as a dependency to the haskell library

@pasqu4le Not sure if I get this right, but you mean we should use the `default_storage.tz` in the test right? Example: `$(fetchValue @FullStorage "out/default_storage.tz" "STORAGE_PATH"))`

dcastro reviewed (approved):
Other than the issues already raised, LGTM :+1:

dcastro commented on ligo/defaults.mligo:
```diff
@@ -0,0 +1,67 @@
+    ; seconds_per_liquidity_outside = 0n
+    ; n_positions = 0n
+    ; fee_growth_outside = {x = 0n; y = 0n}
+    ; sqrt_price = 0n
```
Actually, just noticed some of these fields have different values from `ligo/tests/tests.mligo`. Namely, `n_positions`, `sqrt_price`

rinn7e replied:
Initially, I make it the same as `tests.mligo`, but turn out that it may not be necessary.

I forgot to remove the comment. If you think it is better to use the same value as the `tests.mligo`, I'll make the changes, otherwise I'll remove the comment.

dcastro replied:
I think setting `n_positions` to `1` is necessary, otherwise a user could set a position at `(-inf, +inf)`, then close the position (which would uninitialize the `-inf` and `+inf` ticks), and leave the contract in an irreparable state.

dcastro replied:
I think `sqrt_price` should have the value corresponding to its tick index (using the equation from the spec), if we want positions at `(-inf, +inf)` to work correctly.

rinn7e replied:
Since `sqrt(P) = sqrt(X/Y)`, I'm not quite what would the value of X and Y for -inf and +inf. Do we just set a fixed value for X and Y? Hope to get your advice on this.

dcastro replied:
That's the equation for the _spot price_ (the price at a specific point in time, calculated from the reserves of `x` and `y` at that moment).

The price for a tick `i` is given by:

```
p(i) = exp(0.0001)^i
```

rinn7e replied:
I see, this is still quite confusing to me. I check `tests.mligo` and seems like the formula used is a bit different: `Round[Exp[5/100000*(2^20-1)]*2^80]`. 

I try your formula quickly in a javascript console and got this:
```
Math.exp(0.0001) ^ (10 ^ 100)
111
```
Which seems a bit strange, maybe I get the parenthesis wrong, I try this instead and got
```
Math.exp((0.0001)^ (10^100))
5.92097202766467e+47
```
Since I'm quite lack in this, hope to get your advice on which one is correct.

dcastro replied:
> I  see, this is still quite confusing to me. I check tests.mligo and seems like the formula used is a bit different: `Round[Exp[5/100000*(2^20-1)]*2^80]`.

I ran some calculations, and that equation seems to be equivalent to `p(i) = exp(0.0001)^i` (_taking into account we store the price as a square root * 2^80_).

For example, for the tick `const_max_tick = 1048575`, we should get:

```
sqrt_price = sqrt(exp(0.0001)^1048575) * 2^80
           = 71107673757466966990985105047137336834554167630.0560224...
           ~ 71107673757466966990985105047137336834554167630
```

And for `-const_max_tick = -1048575`, we should get:

```
sqrt_price = sqrt(exp(0.0001)^-1048575) * 2^80
           = 20.5533...
           ~ 21
```

Which matches the values in `ligo/tests/tests.mligo`.

(You can use a calculator [like this one](https://www.mathsisfun.com/calculator-precision.html) to avoid results in scientific notation)

rinn7e replied:
I see, I tried the calculator with `sqrt(exp(0.0001)^(10^100)) * 2^80`. It results in literally infinity. 

Should we use some other value (or just use `1048575` like the test)?

dcastro replied:
Hmm good question, let's resume this conversation in Slack and get other people's opinions on this!

rinn7e replied:
@pasqu4le Maybe you missed this, what do you think about this?

rinn7e replied:
@dcastro I've updated this now, you can take a look again.

pasqu4le commented:
> you mean we should use the default_storage.tz in the test right?

@rinn7e that's right, I think that we should add:
- a function in `Test.SegCFMM.Contract` (like you mentioned)
- a test to check that types match in `Test.SegCFMM.Types`

just like we do for the whole contract

rinn7e commented on Makefile:
```diff
@@ -19,14 +22,20 @@ escape_double_quote = $(subst $\",$\\",$(1))
+	mkdir -p $(OUT)
+
+	$(BUILD_STORAGE) --output-file $(OUT)/default_storage.tz \
+      ligo/main.mligo main "default_storage"
```
I've updated this.

rinn7e commented on ligo/defaults.mligo:
```diff
@@ -0,0 +1,67 @@
+
+
+let default_storage : storage =
+  let infinity : int = pow (10, 100) in
```
Didn't know that we can just put the whole number, I've updated this now.

rinn7e commented:
>     * a function in `Test.SegCFMM.Contract` (like you mentioned)
> 
>     * a test to check that types match in `Test.SegCFMM.Types`

@pasqu4le I've added `fetchValue` for the storage. Not sure how to add the test tho, since to use `evaluateNF_`, I have to add `NFData` instance to all types that are used in `Storage` . Probably not what we want and if the type does not match `fetchValue` will throw errors at compile time. So I added a test like this for now:

```
unit_StorageTypesMatch :: Assertion
unit_StorageTypesMatch = do
  let !(_ :: Storage) = defaultStorage
  pure ()
```
Hope to get your view on this.

pasqu4le commented on haskell/src/Util.hs:
```diff
@@ -4,6 +4,7 @@
 -- | Read a contract at compile time.
 module Util
   ( fetchContract
```
:thinking: in newer morley versions there are utilities that could make us avoid these utility definitions, possibly it's worth updating sooner rather than later.

Since this PR is almost ready however, so I suppose it's better to do so after this is merged.

rinn7e replied:
Yeah, I've updated this as part of this PR (by using `embedContractM` instead): https://github.com/serokell/segmented-cfmm/pull/18#discussion_r700461264

Let me know if this is good enough.

pasqu4le replied:
That's OK, since that PR is based on this one I'll keep any comment related to updating deps for it :ok_hand:

pasqu4le commented on Makefile:
```diff
@@ -66,12 +89,14 @@ metadata: lib all
 test: all
 	$(MAKE) -C haskell test PACKAGE=segmented-cfmm \
-		SEGMENTED_CFMM_PATH=../$(OUT)/segmented_cfmm_default.tz
+		SEGMENTED_CFMM_PATH=../$(OUT)/segmented_cfmm_default.tz \
+		STORAGE_PATH=../$(OUT)/storage_default.tz
```
I'm confused about one thing: here you se these variables to build the storage, however in the `package.yaml` the `extra-source-file` lists only one of them and does so with a (different) hardcoded path:
```yaml
extra-source-files:
- test/segmented_cfmm.tz
```
AFAIK that file should be updated and this one should probably copy the output instead of setting these vars.

rinn7e replied:
The `extra-source-file` is updated as part of this PR: https://github.com/serokell/segmented-cfmm/commit/7b485366ffaecd5fdbeaff3fb1fab10bf86eaad5

I'm not sure where we get the path `test/segmented_cfmm.tz` from. Currently, if I build the project I get this warning: `Warning: File listed in segmented-cfmm.cabal file does not exist: test/segmented_cfmm.tz`.

I try to update the path to $(OUT) instead which is `../out/segmented_cfmm_default.tz`. That results in another warning: 

```Warning: 'extra-source-files: ../out/segmented_cfmm_default.tz' is a relative
path outside of the source tree. This will not work when generating a tarball
with 'sdist'.```

Not sure if that is ideal. So that's why I leave it the way it is. Let me know what you think about this.

pasqu4le replied:
I see, it looks like we mixed two approaches here (env vars and extra-source-file) that don't work all that well together, I'll leave other comments to fix this.

pasqu4le commented on haskell/test/Test/SegCFMM/Types.hs:
```diff
@@ -11,7 +12,14 @@ import Test.HUnit (Assertion)
+unit_ContractTypesMatch :: Assertion
+unit_ContractTypesMatch = evaluateNF_ @(Contract _ _) segCFMMContract
+
+unit_StorageTypesMatch :: Assertion
+unit_StorageTypesMatch = do
+  let !(_ :: Storage) = defaultStorage
+  pure ()
```
@rinn7e I suppose that this is ok, although it doesn't really add anything in term of checks.

IMO once we are using `defaultStorage` in our actual tests we can get rid of this, so for now I'd say to keep this around, but also please add a TODO to remove this once it's no longer useful.

P.s. I suppose that a cleaner way to define this might be:
```suggestion
unit_StorageTypesMatch :: Assertion
unit_StorageTypesMatch = void $ pure (defaultStorage :: Storage)
```

rinn7e replied:
Thanks Pasquale, I updated this.

pasqu4le commented on Makefile:
```diff
@@ -66,12 +89,14 @@ metadata: lib all
 
 test: all
 	$(MAKE) -C haskell test PACKAGE=segmented-cfmm \
```
We can do the same thing here:
```suggestion
test: prepare_lib
	$(MAKE) -C haskell test PACKAGE=segmented-cfmm
```

rinn7e replied:
Updated.

pasqu4le commented on Makefile:
```diff
@@ -66,12 +89,14 @@ metadata: lib all
 test: all
 	$(MAKE) -C haskell test PACKAGE=segmented-cfmm \
-		SEGMENTED_CFMM_PATH=../$(OUT)/segmented_cfmm_default.tz
+		SEGMENTED_CFMM_PATH=../$(OUT)/segmented_cfmm_default.tz \
+		STORAGE_PATH=../$(OUT)/storage_default.tz
```
... and remove the env var settings entirely.
```suggestion
```
(Note: same for the other haskell-related Makefile targets)

How does that sound? Does it seem like a good solution to you?

rinn7e replied:
Seems good to me, I've remove these env vars now.

pasqu4le reviewed (commented):
@rinn7e here's a possible solution to the compiled dependency issue [from here](https://github.com/serokell/segmented-cfmm/pull/12#discussion_r701465600).

Note that if you go with it you'll also need to add the `test/storage_default.tz` in the `extra-source-files` and probably add a copy of the compiled storage in CI as well.

rinn7e commented on Makefile:
```diff
@@ -30,20 +33,40 @@ define build_ligo
+
+$(OUT)/storage_%.tz: ligo/**
+	$(call build_ligo_storage,ligo/main.mligo,$(OUT)/storage_$*.tz,$(LIGO_PRAGMAS))
+
```
Make sense, I've updated this, should be good now.

pasqu4le replied:
@rinn7e the only addition I would make is to remove these files in the `haskell/test` dir on the `clean` target as well

rinn7e replied:
I update the `clean` to remove those files as well.

One problem coming up is that, the compile time code that read contract/storage expect relative path like this:

```
defaultStorage :: Storage
defaultStorage =
  fromVal ($(fetchValue @Storage "../haskell/test/storage_default.tz" "STORAGE_PATH"))
```

This works locally, but won't work in the CI for some reasons: https://buildkite.com/serokell/segmented-cfmm/builds/241#e0c7b80b-0d1f-4f48-a93b-a99d30234584/72-228

As a result, I still have to pass the flag in the Makefile to make the CI passed:
```
test: prepare_lib
	$(MAKE) -C haskell test PACKAGE=segmented-cfmm \
		SEGMENTED_CFMM_PATH=../haskell/test/segmented_cfmm_default.tz \
		STORAGE_PATH=../haskell/test/storage_default.tz
```

Let me know if you're alright with this.

pasqu4le replied:
It should be possible for CI to copy the contract as well.

Firstly I would change the path in `defaultStorage` to be `test/storage_default.tz` which seems like it should work just as well.

Secondly, there's a way to copy the artifacts from one step to another, in the `pipeline.yaml` you can add a utility "env" before the `steps:` definitions:
```yaml
env:
  FETCH_CONTRACT: >-
    mkdir -p haskell/test/
    && buildkite-agent artifact download ligo-out/storage_default.tz haskell/test/
    && buildkite-agent artifact download ligo-out/segmented_cfmm_default.tz haskell/test/
```
and in the haskell steps you can use this in `commands` by calling `eval "$FETCH_CONTRACT"`, e.g.:
```yaml
  - label: build-haskell
    key: build-haskell
    commands:
      - eval "$FETCH_CONTRACT"
      - nix-build ci.nix -A all-components
```

How does it seem to you? Can you try it out?

rinn7e replied:
Seems like the CI only works with path `haskell/test/segmented_cfmm_default.tz`.

I already updated according to your feedback, it manages to download the artifact fine, but still it does not recognized the files: https://buildkite.com/serokell/segmented-cfmm/builds/246#6b59a189-bc36-4695-9fd3-4f68c4b97dda/65-220

pasqu4le replied:
Ah sorry, my suggestion was misguided, I didn't understand your previous comment correctly and only now I noticed that the step didn't depend on `build-ligo`.

It looks like the building is repeated and the files copied already as part of `packages.segmented-cfmm` definition in `ci.nix`, so everything should work without your last commit and without env vars in the Makefile.

I pushed this change [here](https://github.com/serokell/segmented-cfmm/tree/pasqu4le/tcfmm-12-tmp) to check this and as you can see [here](https://buildkite.com/serokell/segmented-cfmm/builds/252#4a0e6beb-8b72-4119-9fd8-b6e42c6d8519) it seems to work fine :shrug: 

Can you check it and cherry-pick it in your own branch?

rinn7e replied:
> It looks like the building is repeated and the files copied already as part of packages.segmented-cfmm definition in ci.nix, so everything should work without your last commit and without env vars in the Makefile.


I see, I've updated `ci.nix` to use `test` dir instead of `haskell/test` dir to match the local build (local build requires `test/<file.tz>` path). Also removed the `eval` path since it is redundant this should be good now, please check this again.

pasqu4le replied:
@rinn7e  LGTM :ok_hand:

rinn7e commented on Makefile:
```diff
@@ -30,20 +33,40 @@ define build_ligo
+	$(call build_ligo_storage,ligo/main.mligo,$(OUT)/storage_$*.tz,$(LIGO_PRAGMAS))
+
+
 lib: all
```
Updated.

pasqu4le reviewed (approved):
one nitpick left. but LGTM

Martoon-00 commented on ligo/defaults.mligo:
```diff
@@ -0,0 +1,53 @@
+
+  let min_tick_state =
+    { prev = { i = -impossible_tick }
+    ; next = { i = int(const_max_tick) }
```
Looks good to me, thanks :+1:
